README.md
<div align="center" style="text-align:center">

# [![icon](https://sv443.net/cdn/jokeapi/icon_readme.png)](#readme)<br>JokeAPI
### A REST API that serves uniformly and well formatted jokes and offers a great variety of filtering methods and response customization
[![GitHub](https://img.shields.io/github/license/Sv443/JokeAPI)](https://sv443.net/LICENSE)
[![Uptime / 7 Days](https://img.shields.io/uptimerobot/ratio/7/m784261094-bff76b959ebb8fc39f7eb2d0)](https://status.sv443.net/)
[![Open Issues](https://img.shields.io/github/issues/Sv443/JokeAPI)](https://github.com/Sv443/JokeAPI/issues)
  
[![Actions Status](https://github.com/Sv443/JokeAPI/workflows/build/badge.svg)](https://github.com/Sv443/JokeAPI/actions)
[![Known Vulnerabilities](https://snyk.io/test/github/Sv443/JokeAPI/badge.svg)](https://snyk.io/test/github/Sv443/JokeAPI)
  
[![Discord](https://img.shields.io/discord/565933531214118942)](https://discord.gg/aBH4uRG)
[![GitHub stars](https://img.shields.io/github/stars/Sv443/JokeAPI?style=social)](https://github.com/Sv443/JokeAPI/stargazers)

<br><br>

# [Documentation](https://jokeapi.dev/)
## [Try it out](https://jokeapi.dev/#try-it) &bull; [Changelog](./changelog.md)


</div>
<br><br>

### Note: due to the different types of humor, JokeAPI contains a wide variety of jokes, of which some are very offensive. They can be filtered out, but use this API at your own risk!
### If you want to contribute to JokeAPI (code, jokes or translations), please read the [contributing guide.](./.github/Contributing.md)

<br><br>

# Official API Wrappers:
- <b><a href="https://github.com/canarado/sv443_jokeapi_wrapper#readme"><img src="./docs/static/external/rust.svg" width="16" height="16"> Rust</a></b> (outdated)
- <b><a href="https://github.com/thenamesweretakenalready/Sv443s-JokeAPI-Python-Wrapper#readme"><img src="./docs/static/external/python.svg" width="16" height="16"> Python</a></b>
- <b><a href="https://github.com/sahithyandev/sv443-joke-api-js-wrapper#readme"><img src="./docs/static/external/nodejs.svg" width="16" height="16"> Node.js</a></b> (incomplete)
- <b><a href="https://github.com/Icelain/jokeapi#readme"><img src="./docs/static/external/golang.svg" width="16" height="16"> Go</a></b>


<br>

# Projects (I'm aware of) that use JokeAPI:
- [dozens Advices](https://github.com/ZephyrVentum/dozens-Advices) by [ZephyrVentum](https://github.com/ZephyrVentum)
- [Shadow-bot](https://github.com/QGIsK/Shadow-bot) by [QGIsK](https://github.com/QGIsK)
- [Jokepy](https://github.com/aksty/Jokepy) by [aksty](https://github.com/aksty)
- [Dark1](https://github.com/whiteadi/Dark1) by [whiteadi](https://github.com/whiteadi)
- [Prejudice Networks](https://github.com/LiamTownsley/Prejudice-Networks) by [Liam Townsley](https://github.com/LiamTownsley)
- [https://irshad.ml/humour.html](https://irshad.ml/humour.html) by [draco-malfoy](https://github.com/draco-malfoy)
- [Random Joke Generator with Flutter](https://github.com/variousnabil/Random-Joke-Generator-with-Flutter) by [variousnabil](https://github.com/variousnabil)
- [comma](https://thatcopy.pw/comma) by [ThatCopy](https://github.com/ThatCopy)
- [Joke Teller](https://github.com/AlHood77/Joke_Teller) by [AlHood77](https://github.com/AlHood77)
- [jokes](https://github.com/remcoov/jokes) by [remcoov](https://github.com/remcoov)
  
[Submit an issue](https://github.com/Sv443/JokeAPI/issues/new/choose) and select `New Dependent` to get your project added here


<br><br>
<div align="center" style="text-align:center">

Made with ❤️ by [Sv443](https://github.com/Sv443)  
If you want to, you can support me on [Patreon](https://patreon.com/Sv443_), [Ko-fi](https://ko-fi.com/Sv443_) or [PayPal](https://paypal.me/SvenFehler)

</div>


CHANGELOG.md
# JokeAPI Changelog (Version 2.2.2)

## [Planned for future releases]
- Allow definition of max requests per minute per each client ([issue #37](https://github.com/Sv443/JokeAPI/issues/37))
- Add positive flags and a "?whitelistFlags" param ([issue #127](https://github.com/Sv443/JokeAPI/issues/127))
- Add Unit Tests ([issue #121](https://github.com/Sv443/JokeAPI/issues/121))
- Serve docs with nginx to speed up page load times ([issue #118](https://github.com/Sv443/JokeAPI/issues/118))


## [CURRENT: 2.2.2] - The Seasonal Update
- Added some new seasonal categories ([issue #180](https://github.com/Sv443/JokeAPI/issues/180))
    - Spooky (Halloween)
    - Christmas (thanks for the suggestion Dan)
- Added dependent ToastIT-dev/PoshBot.Joker ([issue #173](https://github.com/Sv443/JokeAPI/issues/173))
- Added a bunch of joke submissions
- Reliability improvements in the documentation ([issue #181](https://github.com/Sv443/JokeAPI/issues/181))
- Added POST as a method to submit data to the API since it makes more sense than PUT (maybe deprecating that some time)


## [2.2.1] - Version 2.2 Hotfix
- Added the Czech translation (PR #137) - provided by @ThatCopy (https://github.com/ThatCopy) - thanks :)
- Added a missing conversion mapping when using format `txt` and the `amount` parameter on endpoint `/joke` ([issue #138](https://github.com/Sv443/JokeAPI/issues/138))
- Fixed wrong URLs in the documentation's JavaScript
- Fixed crash when a client asks for a translation that doesn't exist
- Slightly improved the documentation
- Applied a few security patches


## [2.2.0]
- Added joke category "Pun" ([issue #105](https://github.com/Sv443/JokeAPI/issues/105))
- Added "?amount" parameter to joke endpoint so multiple jokes can be fetched at once ([issue #126](https://github.com/Sv443/JokeAPI/issues/126))
- Added support for jokes and error messages of different languages ([issue #75](https://github.com/Sv443/JokeAPI/issues/75))
    - /langcode/{LANGUAGE} endpoint
    - /languages endpoint
    - "?lang=code" URL parameter
- Fixed ID caching (again, sigh) ([issue #80](https://github.com/Sv443/JokeAPI/issues/80))
- Added pm2 custom metrics ([issue #91](https://github.com/Sv443/JokeAPI/issues/91))
- Fixed HTTP 403 errors ([issue #96](https://github.com/Sv443/JokeAPI/issues/96))
- Remade the URL parser using a package ([issue #97](https://github.com/Sv443/JokeAPI/issues/97))
- Daemonized the API token refreshing ([issue #102](https://github.com/Sv443/JokeAPI/issues/102))
- Rate limiting joke submissions more harshly now ([issue #104](https://github.com/Sv443/JokeAPI/issues/104))
- Fixed error where the end of the payload were cut off, thus invalidating JSON ([issue #119](https://github.com/Sv443/JokeAPI/issues/119))
- Joke submission property order is now enforced, improving uniformity ([issue #120](https://github.com/Sv443/JokeAPI/issues/120))
- Joke submissions are now validated to make sure they don't contain fancy Unicode chars ([issue #123](https://github.com/Sv443/JokeAPI/issues/123))


## [2.1.5] - 2020 Q3 general patch #2
- Ditched my botched rate limiting package for a "commercial" one ([issue #113](https://github.com/Sv443/JokeAPI/issues/113))
- Added API token section to documentation ([issue #114](https://github.com/Sv443/JokeAPI/issues/114))
- Client now receives a "Token-Valid" header with the value 0 or 1 depending on token validity ([issue #115](https://github.com/Sv443/JokeAPI/issues/115))
- Renamed "X-Auth-Token" header to "Authorization" so requests don't get blocked by Cloudflare ([issue #117](https://github.com/Sv443/JokeAPI/issues/117))
- Cleaned up a lot of code


## [2.1.4] - 2020 Q3 general patch #1
- Fixed the IP getter module for like the 500th time now


## [2.1.3] - 2020 Q2 general patch #1
- Added option to disable all console output but error messages (GitHub [issue #72](https://github.com/Sv443/JokeAPI/issues/72))
- The content of jokes in the joke submission form is now correctly escaped and can no longer mess up the page (GitHub [issue #68](https://github.com/Sv443/JokeAPI/issues/68))
- Fixed crash when parsing a malformatted URI (GitHub [issue #69](https://github.com/Sv443/JokeAPI/issues/69) (nice))
- Re-flagged some jokes
- Updated dependencies


## [2.1.2] - Plain Text update
- Added file format "txt" to receive data as plain text
- Fixed the joke submission URL in the /info endpoint data
- Added HTTP error codes:
    - 413 Payload Too Large
    - 414 URI Too Long


## [2.1.1] - Auth update hotfix
- Fixed incorrect error cause when using an out-of-range ID range parameter (see GitHub [issue #54](https://github.com/Sv443/JokeAPI/issues/54))
- Added submission form (https://sv443.net/jokeapi/v2#submit)


## [2.1.0] - The auth update
- Added an authorization header to make whitelisting possible without needing to have a static IP
    - Added the script "npm run add-token [amount]" to add one or more tokens
- Improved the documentation (see GitHub [issue #52](https://github.com/Sv443/JokeAPI/issues/52))
- Fixed a few bugs in the documentation


## [2.0.1] - A few hotfixes for the big 2.0.0 updates and some very very minor features I wanted to add
- Hotfixed a few bugs from the big 2.0.0 update
- Fixed joke ID caching (to not serve the same jokes multiple times)
- Added three new commands that are run through CI and before contributing
- Enabled automated code fixing with ESLint


## [2.0.0] - The complete rewrite - JokeAPI was completely rewritten and should now run like 100x better and be more easy to develop and maintain
- Massively improved the "Try it out" section in the docs
- Reformatted the jokes to always contain all flags
- Added support for selecting multiple categories at once (for example: "https://sv443.net/jokeapi/v2/joke/Dark,Miscellaneous/")
- Massive performance improvements:

settings.js
const packageJSON = require("./package.json");
const jsl = require("svjsl");
const col = jsl.colors.fg;
const bgc = jsl.colors.bg;

const settings = {
    debug: {
        verboseLogging: true,      // set to true to enable extra debug output
        progressBarDisabled: true,  // set to true to disable the progress bar - greatly improves readability of verbose debug output
        onlyLogErrors: true,        // set to true to disable sending any console logs but error messages
    },
    info: {
        name: "JokeAPI",                                // the name of JokeAPI
        desc: packageJSON.description,                  // the description of JokeAPI
        projGitHub: "https://github.com/Sv443/JokeAPI", // URL to the project's GitHub page
        version: packageJSON.version,                   // the version as a string
        versionInt: packageJSON.version.split("."),     // the version as a number array
        docsURL: packageJSON.homepage,                  // the URL to the documentation of JokeAPI
        author: {
            name: packageJSON.author.name,   // author name
            email: packageJSON.author.email, // author email
            website: packageJSON.author.url, // author website
            github: `https://github.com/${packageJSON.author.name}`, // author github page
        },
        privacyPolicyUrl: "https://sv443.net/privacypolicy/en"
    },
    wrapper: {
        mainFilePath: "./src/main.js",          // main script file
        skipWrapping: true,                    // whether or not to skip the wrapping through node-wrap
        wrapperSettings: {
            console: true,                      // whether Node-Wrap should log to the console
            crashTimeout: 2000,                 // timeout (in ms) until the process should be restarted after a crash
            logFile: "./data/logs/wrapper.log", // Node-Wrap log file
            logTimestamp: true,                 // whether to add a timestamp to the log
            restartOnCrash: true,               // whether to restart the process after a crash
            restartTimeout: 0,                  // timeout (in ms) until the process should be started again after a restart has been requested
        },
    },
    init: {
        initDirs: [ // directories that should be generated if they don't exist - paths relative to root of project - doesn't necessarily need trailing slash
            "./data/logs",
            "./data/submissions",
            "./docs/compiled",
            "./data/lists"
        ],
        exitSignals: [ // all signals that should cause a soft exit
            "SIGINT",
            "SIGTERM"
        ],
    },
    logging: {
        logChar: "▌",                  // character that gets logged on each request
        spacerAfter: 10,               // after how many logged requests a spacer should be put - set to 0 to disable
        disableLogging: false,         // set to true to disable logging a character on each request
        blacklistLoggingEnabled: true, // whether or not to log the character when an IP is on the blacklist
    },
    jokes: {
        jokesFormatVersion: 3,                               // current joke format version
        jokesFolderPath: "./data/jokes/",                    // path to the jokes folder - needs trailing slash
        jokeSubmissionURL: `${packageJSON.homepage}#submit`, // joke submission url
        jokeSubmissionPath: "./data/submissions/",           // path to a directory where joke submissions should be saved to - needs trailing slash
        submissions: {
            timeFrame: 60,                              // time frame of submission rate limiter (in seconds)
            rateLimiting: 3,                            // how many requests per timeframe should be allowed
            invalidCharRegex: /(?![\u0000-\u0fff])./gm, // eslint-disable-line no-control-regex
        },
        jokesTemplateFile: "template.json",  // relative to "jokes.jokesFolderPath"
        possible: {
            anyCategoryName: "Any", // the name of the "Any" category - readable name
            categories: [           // all categories (excluding "Any") - case insensitive / readable name
                "Misc",
                "Programming",
                "Dark",
                "Pun",
                "Spooky",
                "Christmas"
            ],
            categoryAliases: { // aliases of categories. Key gets auto-converted to value. Value has to be present in the "categories" array above!
                "Miscellaneous": "Misc",
                "Coding": "Programming",
                "Development": "Programming",
                "Halloween": "Spooky"
            },
            flags: [ // all flags - HAVE TO BE LOWER CASE!
                "nsfw",
                "religious",
                "political",
                "racist",
                "sexist",
                "explicit",
            ],
            formats: [ // all file formats - HAVE TO BE LOWER CASE!
                "json",
                "xml",
                "yaml",
                "txt",
            ],
            types: [ // all joke types - HAVE TO BE LOWER CASE!
                "single",
                "twopart"
            ],
        },
        fileFormatsPath: "./data/fileFormats.json", // path to the file formats file
        defaultFileFormat: {
            fileFormat: "json",           // the default file format string
            mimeType: "application/json", // the default file format mime type
        },
        lastIDsMaxLength: 15,          // the maximum amount of joke IDs that get saved to the blacklist-array
        jokeRandomizationAttempts: 25, // after how many attempts of selecting a random joke to stop trying
        splitChars: [",", "+", "-"],   // which characters should separate the values of parameters with support for multiple values
        splitCharRegex: /[,+-]/gm,     // which characters should separate the values of parameters with support for multiple values
        maxAmount: 10,                 // the maximum amount of jokes that can be fetched with a single call to the get jokes endpoint
        encodeAmount: 5,               // if more than this number of jokes is requested, encode them
    },
    httpServer: {
        port: 8076,           // http server port
        allowCORS: true,      // whether or not to allow Cross Origin Resource Sharing
        rateLimiting: 60,     // amount of allowed requests per below defined timeframe
        timeFrame: 60,        // timeframe in seconds
        urlPathOffset: 0,     // example: "/jokeapi/info" with an offset of 1 will only start parsing the path beginning at "info" - an Apache reverse proxy will do this automatically though
        maxPayloadSize: 5120, // max size (in bytes) that will be accepted in a PUT request - if payload exceeds this size, it will abort with status 413
        maxUrlLength: 250,    // max amount of characters of the URL - if the URL is longer than this, the request will abort with status 414
        disableCache: true,   // whether or not to disable the cache - default: true (setting to false may prevent the users from getting new jokes)
        infoHeaders: true,    // whether or not to add an informational header about JokeAPI to each request
        reverseProxy: true,   // whether or not JokeAPI gets its requests from a reverse proxy
        ssl: {
            enabled: false,                // whether SSL is enabled
            certFile: "./.ssl/cert-xy.pem" // to be implemented (issue #185)
        },
        ipSanitization: {     // used to sanitize IP addresses so they can be used in file paths
            regex: /[^A-Za-z0-9\-_./]|^COM[0-9]([/.]|$)|^LPT[0-9]([/.]|$)|^PRN([/.]|$)|^CLOCK\$([/.]|$)|^AUX([/.]|$)|^NUL([/.]|$)|^CON([/.]|$)/gm,
            replaceChar: "#", // what character to use instead of illegal characters
        },
        ipHashing: {
            enabled: true,       // hashes all IP addresses. If set to false, JokeAPI is not GDPR compliant anymore!
            algorithm: "sha256", // the algorithm of the hash - available algorithms depend on the OpenSSL version installed on the machine (on linux can be listed with "openssl list -digest-algorithms")
            digest: "hex",       // the output format of the hash - can be "base64", "hex" or "latin1"
        },
        encodings: {
            gzip: true,    // Whether or not Gzip encoding should be enabled for the documentation page
            deflate: true, // Whether or not Deflate encoding should be enabled for the documentation page
            brotli: true,  // Whether or not Brotli encoding should be enabled for the documentation page
        },
        encodingPriority: [ // The priority of the encodings. Items with a lower array index (further to the left) have a higher priority
            "brotli", "gzip", "deflate"
        ],
    },
    errors: {
        errorLogDir: "./data/logs/",               // path to the error log directory - needs trailing slash
        errorMessagesPath: "./data/errorMessages", // path to error messages file
    },
    lists: {
        blacklistPath: "./data/lists/ipBlacklist.json",             // path to the IP blacklist
        whitelistPath: "./data/lists/ipWhitelist.json",             // path to the IP whitelist
        consoleBlacklistPath: "./data/lists/consoleBlacklist.json", // path to the IP console blacklist
    },
    documentation: {
        dirPath: "./docs/",                // path to the documentation directory - needs trailing slash
        compiledPath: "./docs/compiled/", // path to the compiled docs directory - needs trailing slash
        faviconPath: "./docs/static/favicon.ico", // path to the favicon.ico file - don't add trailing slash
        rawDirPath: "./docs/raw/",            // path to the raw documentation files directory - needs trailing slash
        daemonInterval: 2,                    // interval (in seconds) at which the daemon checks for changes in the documentation directory
        errorPagePath: "./docs/raw/errorPage.html", // path to the error page
        codeFontFileName: "static/external/CascadiaCode-Regular-VTT_1911.21.ttf", // the name of the font file that is going to be used in code blocks - has to be in the directory specified with the above property "dirPath"
        submissionForm: {
            dirPath: "./docs/raw/", // path to the submission form directory - needs trailing slash
            fileNames: {
                html: "submit.html", // name of the HTML file of the submission form - relative to the parameter "documentation.submissionForm.dirPath"
                js: "../static/submit.js", // name of the JS file of the submission form - relative to the parameter "documentation.submissionForm.dirPath"
                css: "../static/submit.css", // name of the CSS file of the submission form - relative to the parameter "documentation.submissionForm.dirPath"
            },
        },
    },
    endpoints: {
        dirPath: "./endpoints/", // path to the dir containing all the endpoint scripts
        ratelimitBlacklist: [    // calling an endpoint in this array will not count towards the rate limit counter
            "static",
        ],
    },
    colors: {
        success: col.green,     // when request was successful
        error: col.red,         // when request was errored
        ratelimit: col.magenta, // when request was rate limited
        docs: col.yellow,                      // when docs were requested
        blacklisted: bgc.red + col.yellow,     // when a request IP is blacklisted
        docsrecompiled: bgc.yellow + col.blue, // when the docs were recompiled
    },
    analytics: {
        enabled: false, // whether or not the analytics module should be enabled
        dirPath: "./data/analytics/", // path to the analytics directory - needs trailing slash
        sqlTableName: "analytics",    // name of the SQL table
    },
    sql: { // (login credentials are set in the .env file)
        host: "localhost",   // IP address to the DB host - default for local device is "localhost"
        database: "jokeapi", // the name of the DB
        port: 3306,          // the port of the DB - default is 3306
    },
    auth: {
        tokenListFile: "./data/tokens.json", // path to the token list file
        tokenHeaderName: "authorization",     // the name of the token header (lower case)
        tokenValidHeader: "Token-Valid",     // the name of the token validity response header (normal case, not lower case)
        daemonInterval: 20, // after how many seconds the auth tokens should be refreshed
    },
    languages: {
        langFilePath: "./data/languages.json",        // file containing all language codes and corresponding language information
        defaultLanguage: "en",                        // default language (two character code, lowercase)
        translationsFile: "./data/translations.json", // translations file
    },
    tests: { // unit tests
        location: "./tests/",  // folder where unit tests are located - requires trailing slash
        initPingInterval: 250, // in ms - interval between init pings (default: 250)
    }
}

module.exports = Object.freeze(settings); // use Object.freeze() to prevent modifications at runtime
